com.unity.ml-agents.extensions/Runtime/MultiAgent/BaseMultiAgentGroup.cs

```diff
@@ -0,0 +1,144 @@
+using System;
+using System.Linq;
+using System.Collections.Generic;
+
+namespace Unity.MLAgents.Extensions.MultiAgent
+{
+    /// <summary>
+    /// A base class implementation of MultiAgentGroup.
+    /// </summary>
+    public class BaseMultiAgentGroup : IMultiAgentGroup, IDisposable
+    {
+        int m_StepCount;
+        int m_GroupMaxStep;
+        readonly int m_Id = MultiAgentGroupIdCounter.GetGroupId();
+        HashSet<Agent> m_Agents = new HashSet<Agent>();
+
+
+        public void Dispose()
+        {
+            while (m_Agents.Count > 0)
+            {
+                UnregisterAgent(m_Agents.First());
+            }
+        }
+
+        /// <inheritdoc />
+        public virtual void RegisterAgent(Agent agent)
+        {
+            if (!m_Agents.Contains(agent))
+            {
+                agent.SetMultiAgentGroup(this);
+                m_Agents.Add(agent);
+                agent.UnregisterFromGroup += UnregisterAgent;
```

Comment:
If we would like the group to constantly check on agents instead of agent triggering it, it will probably needs to be a monobehaviour to do this check in every FixedUpdate (or passively check it when any method is called which sounds buggy). Also this doesn't work if the agent is still alive but assigned to another group.